File: src/typegraph/core/src/t.rs
// Copyright Metatype OÜ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

use crate::errors::Result;
use crate::errors::TgError;
use crate::sdk::core::{
    Handler, TypeEither, TypeFloat, TypeFunc, TypeInteger, TypeList, TypeOptional, TypeString,
    TypeStruct, TypeUnion,
};
use crate::types::RefAttr;
use crate::types::TypeRefBuilder;
use crate::types::{Named as _, TypeId, TypeRef};

#[cfg(test)]
use common::typegraph::{Injection, InjectionData, SingleValue};

pub trait TypeBuilder {
    fn build(&self) -> Result<TypeId>;

    fn optional(&self) -> Result<OptionalBuilder> {
        Ok(optional(self.build()?))
    }

    fn build_named(&self, name: impl Into<String>) -> Result<TypeId> {
        Ok(self.build()?.named(name)?.id())
    }

    #[cfg(test)]
    fn as_id(&self) -> Result<TypeRef> {
        use crate::types::type_ref::AsId;
        self.build()?.as_id(false)
    }

    #[cfg(test)]
    fn inject(&self, injection: Injection) -> Result<TypeRef> {
        use crate::types::type_ref::WithInjection;
        self.build()?.with_injection(injection)
    }

    #[cfg(test)]
    fn set_value<V: serde::ser::Serialize>(&mut self, val: V) -> Result<TypeRef> {
        self.inject(Injection::Static(InjectionData::SingleValue(SingleValue {
            value: serde_json::to_value(&val).unwrap(),
        })))
    }

    #[cfg(test)]
    fn config(
        self,
        key: impl Into<String>,
        value: impl serde::ser::Serialize,
    ) -> Result<TypeRefBuilder>
    where
        Self: Sized,
    {
        Ok(TypeRef::from_type(
            self.build()?.as_type()?,
            RefAttr::runtime("", serde_json::json!({key.into(): value})),
        ))
    }
}

impl<T> TypeBuilder for &mut T
where
    T: TypeBuilder,
{
    fn build(&self) -> Result<TypeId> {
        (**self).build()
    }
}

impl TypeBuilder for TypeId {
    fn build(&self) -> Result<TypeId> {
        Ok(*self)
    }
}

impl TypeBuilder for TypeRef {
    fn build(&self) -> Result<TypeId> {
        Ok(self.id())
    }
}

impl<T> TypeBuilder for Result<T>
where
    T: TypeBuilder,
{
    fn build(&self) -> Result<TypeId, TgError> {
        self.as_ref().map_err(|e| e.clone())?.build()
    }
}

#[derive(Default)]
pub struct BooleanBuilder;

pub fn boolean() -> BooleanBuilder {
    Default::default()
}

#[derive(Default)]
pub struct IntegerBuilder {
    data: TypeInteger,
}

#[allow(clippy::derivable_impls)]
impl Default for TypeInteger {
    fn default() -> Self {
        Self {
            min: None,
            max: None,
            exclusive_minimum: None,
            exclusive_maximum: None,
            multiple_of: None,
            enumeration: None,
        }
    }
}

impl IntegerBuilder {
    #[allow(dead_code)]
    pub fn min(mut self, min: i32) -> Self {
        self.data.min = Some(min);
        self
    }

    #[allow(dead_code)]
    pub fn max(mut self, max: i32) -> Self {
        self.data.max = Some(max);
        self
    }

    #[allow(dead_code)]
    pub fn x_min(mut self, min: i32) -> Self {
        self.data.exclusive_minimum = Some(min);
        self
    }

    #[allow(dead_code)]
    pub fn x_max(mut self, max: i32) -> Self {
        self.data.exclusive_maximum = Some(max);
        self
    }
}

pub fn integer() -> IntegerBuilder {
    Default::default()
}

#[derive(Default)]
pub struct FloatBuilder {
    data: TypeFloat,
}

#[allow(clippy::derivable_impls)]
impl Default for TypeFloat {
    fn default() -> Self {
        Self {
            min: None,
            max: None,
            exclusive_minimum: None,
            exclusive_maximum: None,
            multiple_of: None,
            enumeration: None,
        }
    }
}

impl FloatBuilder {
    #[allow(dead_code)]
    pub fn min(mut self, min: f64) -> Self {
        self.data.min = Some(min);
        self
    }

    #[allow(dead_code)]
    pub fn max(mut self, max: f64) -> Self {
        self.data.max = Some(max);
        self
    }

    #[allow(dead_code)]
    pub fn x_min(mut self, min: f64) -> Self {
        self.data.exclusive_minimum = Some(min);
        self
    }

    #[allow(dead_code)]
    pub fn x_max(mut self, max: f64) -> Self {
        self.data.exclusive_maximum = Some(max);
        self
    }
}

pub fn float() -> FloatBuilder {
    Default::default()
}

#[derive(Default)]
pub struct StringBuilder {
    data: TypeString,
}

#[allow(clippy::derivable_impls)]
impl Default for TypeString {
    fn default() -> Self {
        Self {
            min: None,
            max: None,
            format: None,
            pattern: None,
            enumeration: None,
        }
    }
}

pub fn string() -> StringBuilder {
    Default::default()
}

impl StringBuilder {
    #[allow(dead_code)]
    pub fn format(&mut self, format: impl Into<String>) -> &mut Self {
        self.data.format = Some(format.into());
        self
    }

    pub fn enum_(&mut self, values: Vec<String>) -> &mut Self {
        self.data.enumeration = Some(
            values
                .into_iter()
                .map(|v| serde_json::to_string(&serde_json::Value::String(v)).unwrap())
                .collect(),
        );
        self
    }
}

#[derive(Default)]
pub struct OptionalBuilder {
    data: TypeOptional,
}

impl Default for TypeOptional {
    fn default() -> Self {
        Self {
            of: u32::MAX,
            default_item: None,
        }
    }
}

pub fn optional(ty: TypeId) -> OptionalBuilder {
    OptionalBuilder {
        data: TypeOptional {
            of: ty.into(),
            default_item: None,
        },
    }
}

pub fn optionalx(item_builder: impl TypeBuilder) -> Result<OptionalBuilder> {
    Ok(optional(item_builder.build()?))
}

#[derive(Default)]
pub struct ListBuilder {
    data: TypeList,
}

impl Default for TypeList {
    fn default() -> Self {
        Self {
            of: u32::MAX,
            min: None,
            max: None,
            unique_items: None,
        }
    }
}

pub fn list(ty: TypeId) -> ListBuilder {
    ListBuilder {
        data: TypeList {
            of: ty.into(),
            ..Default::default()
        },
    }
}

pub fn listx(item_builder: impl TypeBuilder) -> Result<ListBuilder> {
    Ok(list(item_builder.build()?))
}

#[derive(Default)]
pub struct UnionBuilder {
    data: TypeUnion,
}

#[allow(clippy::derivable_impls)]
impl Default for TypeUnion {
    fn default() -> Self {
        Self {
            variants: Default::default(),
        }
    }
}

impl UnionBuilder {
    pub fn add(&mut self, ty: TypeId) -> &mut Self {
        self.data.variants.push(ty.0);
        self
    }

    pub fn addx(&mut self, ty: impl TypeBuilder) -> Result<&mut Self> {
        self.add(ty.build()?);
        Ok(self)
    }
}

pub fn union(variants: impl IntoIterator<Item = TypeId>) -> UnionBuilder {
    UnionBuilder {
        data: TypeUnion {
            variants: variants.into_iter().map(|tid| tid.0).collect(),
        },
    }
}

macro_rules! unionx {
    [ $($ty:expr),* ] => {
        $crate::t::union(vec![$($ty.build()?),*])
    };

    [ $($ty:expr),*, ] => {
        crate::t::unionx![$($ty),*]
    };
}
pub(crate) use unionx;

#[derive(Default)]
pub struct EitherBuilder {
    data: TypeEither,
}

#[allow(clippy::derivable_impls)]
impl Default for TypeEither {
    fn default() -> Self {
        Self {
            variants: Default::default(),
        }
    }
}

pub fn either(variants: impl IntoIterator<Item = TypeId>) -> EitherBuilder {
    EitherBuilder {
        data: TypeEither {
            variants: variants.into_iter().map(|tid| tid.0).collect(),
        },
    }
}

macro_rules! eitherx {
    [ $($ty:expr),* ] => {
        $crate::t::either(vec![$($ty.build()?),*])
    };

    [ $($ty:expr),*, ] => {
        crate::t::eitherx![$($ty),*]
    };
}
pub(crate) use eitherx;

#[derive(Default)]
pub struct StructBuilder {
    data: TypeStruct,
}

#[allow(clippy::derivable_impls)]
impl Default for TypeStruct {
    fn default() -> Self {
        Self {
            props: Vec::new(),
            additional_props: false,
            min: None,
            max: None,
            enumeration: None,
        }
    }
}

pub fn struct_() -> StructBuilder {
    Default::default()
}

pub fn struct_extends(ty: TypeId) -> Result<StructBuilder> {
    Ok(StructBuilder {
        data: TypeStruct {
            props: ty.as_struct().map(|typ| typ.data.props.clone())?,
            ..Default::default()
        },
    })
}

impl StructBuilder {
    pub fn prop(&mut self, name: impl Into<String>, ty: TypeId) -> &mut Self {
        self.data.props.push((name.into(), ty.into()));
        self
    }

    pub fn propx(
        &mut self,
        name: impl Into<String>,
        builder: impl TypeBuilder,
    ) -> Result<&mut Self> {
        self.data.props.push((name.into(), builder.build()?.into()));
        Ok(self)
    }

    #[allow(dead_code)]
    pub fn props(&mut self, props: impl IntoIterator<Item = (String, TypeId)>) {
        self.data
            .props
            .extend(props.into_iter().map(|(name, ty)| (name, ty.into())));
    }

    pub fn min(&mut self, min: u32) -> &mut Self {
        self.data.min = Some(min);
        self
    }

    #[allow(dead_code)]
    pub fn max(&mut self, max: u32) -> &mut Self {
        self.data.max = Some(max);
        self
    }
}

#[derive(Default)]
pub struct FuncBuilder {
    data: TypeFunc,
}

impl Default for TypeFunc {
    fn default() -> Self {
        Self {
            inp: u32::MAX,
            out: u32::MAX,
            parameter_transform: None,
            mat: u32::MAX,
            rate_calls: false,
            rate_weight: None,
        }
    }
}

#[allow(dead_code)]
pub fn func(inp: TypeId, out: TypeId, mat: u32) -> Result<TypeId> {
    FuncBuilder {
        data: TypeFunc {
            inp: inp.into(),
            out: out.into(),
            mat,
            ..Default::default()
        },
    }
    .build()
}

pub struct RefBuilder {
    name: String,
    attribute: Option<RefAttr>,
}

pub fn ref_(name: impl Into<String>, attribute: Option<RefAttr>) -> RefBuilder {
    RefBuilder {
        name: name.into(),
        attribute,
    }
}

macro_rules! impl_type_builder {
    ( $ty:ty, $build:ident ) => {
        impl TypeBuilder for $ty {
            fn build(&self) -> Result<TypeId> {
                let res = $crate::Lib::$build(self.data.clone())?;
                Ok(res.into())
            }
        }
    };

    ( $ty:ty, $build:ident, true ) => {
        impl TypeBuilder for $ty {
            fn build(&self) -> Result<TypeId> {
                let builder = self.clone();
                Ok($crate::Lib::$build(builder.data.clone())?.into())
            }
        }
    };
}
impl TypeBuilder for BooleanBuilder {
    fn build(&self) -> Result<TypeId> {
        let res = crate::Lib::booleanb()?;
        Ok(res.into())
    }
}

impl_type_builder!(IntegerBuilder, integerb);
impl_type_builder!(FloatBuilder, floatb);
impl_type_builder!(OptionalBuilder, optionalb);
impl_type_builder!(StringBuilder, stringb);
impl_type_builder!(ListBuilder, listb);
impl_type_builder!(UnionBuilder, unionb);
impl_type_builder!(EitherBuilder, eitherb);
impl_type_builder!(StructBuilder, structb);
impl_type_builder!(FuncBuilder, funcb, true);

impl TypeBuilder for RefBuilder {
    fn build(&self) -> Result<TypeId> {
        Ok(crate::Lib::refb(
            self.name.clone(),
            self.attribute
                .as_ref()
                .map(|attr| serde_json::to_string(&attr).unwrap()),
        )?
        .into())
    }
}

impl TypeBuilder for TypeRefBuilder {
    fn build(&self) -> Result<TypeId> {
        (*self).clone().register().map(|r| r.id())
    }
}

File: src/typegraph/python/typegraph/t.py
# Copyright Metatype OÜ, licensed under the Mozilla Public License Version 2.0.
# SPDX-License-Identifier: MPL-2.0

import json as JsonLib
from typing import Any, Dict, List, Optional, Tuple, Union, Literal
import copy

from typing_extensions import Self

from typegraph.effects import EffectType
from typegraph.gen.exports.core import (
    FuncParams,
    ParameterTransform,
    TypeList,
    TypeEither,
    TypeFile,
    TypeFloat,
    TypeFunc,
    TypeInteger,
    TypeOptional,
    TypeString,
    TypeStruct,
    TypeUnion,
    PolicySpec as WitPolicySpec,
)
from typegraph.gen.exports.runtimes import EffectRead
from typegraph.gen.types import Err
from typegraph.graph.typegraph import (
    ErrorStack,
    core,
    store,
    ApplyFromArg,
    ApplyFromContext,
    ApplyFromParent,
    ApplyFromSecret,
    ApplyFromStatic,
)
from typegraph.io import Log
from typegraph.injection import (
    serialize_generic_injection,
    serialize_parent_injection,
    serialize_static_injection,
)
from typegraph.policy import Policy, PolicyPerEffect, PolicySpec, get_policy_chain
from typegraph.runtimes.deno import Materializer
from typegraph.utils import (
    ConfigSpec,
    build_reduce_entries,
    serialize_config,
)
from typegraph.wit import wit_utils

# TODO: better approach?
og_list = list
og_float = float


class typedef:
    _id: int
    injection: Optional[str]
    policy_chain: Optional[List[WitPolicySpec]]
    name: Optional[str]

    def __init__(self, id: int):
        self._id = id
        self.injection = None
        self.policy_chain = None
        self.name = None

    def __repr__(self):
        res = core.get_type_repr(store, self._id)
        if isinstance(res, Err):
            raise ErrorStack(res.value)
        return res.value

    def with_policy(self, *policies: Optional[PolicySpec]) -> Self:
        policy_chain = get_policy_chain(policies)
        res = core.with_policy(store, self._id, policy_chain)
        if isinstance(res, Err):
            raise ErrorStack(res.value)

        ret = copy.copy(self)
        ret._id = res.value
        ret.policy_chain = policy_chain
        return ret

    def rename(self, name: str) -> Self:
        res = core.rename_type(store, self._id, name)

        if isinstance(res, Err):
            raise ErrorStack(res.value)

        ret = copy.copy(self)
        ret._id = res.value
        ret.name = name
        return ret

    def _with_injection(self, injection: str) -> Self:
        res = core.with_injection(store, self._id, injection)
        if isinstance(res, Err):
            raise ErrorStack(res.value)

        ret = copy.copy(self)
        ret._id = res.value
        ret.injection = injection
        return ret

    def optional(
        self,
        default_value: Optional[Any] = None,
        config: Optional[ConfigSpec] = None,
    ) -> "optional":
        if isinstance(self, optional):
            return self
        return optional(self, default_item=default_value, config=config)

    def set(self, value: Union[any, Dict[EffectType, any]]):
        return self._with_injection(serialize_static_injection(value))

    def inject(self, value: Union[any, Dict[EffectType, any]]):
        return self._with_injection(serialize_generic_injection("dynamic", value))

    def from_context(self, value: Union[str, Dict[EffectType, str]]):
        return self._with_injection(serialize_generic_injection("context", value))

    def from_secret(self, value: Union[str, Dict[EffectType, str]]):
        return self._with_injection(serialize_generic_injection("secret", value))

    def from_parent(self, value: Union[str, Dict[EffectType, str]]):
        return self._with_injection(serialize_parent_injection(value))

    def from_random(self):
        return self._with_injection(serialize_generic_injection("random", None))


class _TypeWithPolicy(typedef):
    base: "typedef"
    policy: Tuple[Union[Policy, PolicyPerEffect], ...]

    def __init__(
        self,
        id: int,
        base: "typedef",
        policy: Tuple[Union[Policy, PolicyPerEffect], ...],
    ):
        super().__init__(id)
        self.base = base
        self.policy = policy

    def __getattr__(self, name):
        if name == "policy":
            return self.policy
        return getattr(self.base, name)


# self._id refer to the wrapper id
# self.* refer to the base id
class _TypeWrapper(typedef):
    base: "typedef"

    def __init__(
        self,
        id: int,
        base: "typedef",
    ):
        super().__init__(id)
        self.base = base

    def __getattr__(self, name):
        if name == "id":
            return self._id
        return getattr(self.base, name)


AsId = Union[bool, Literal["simple", "composite"]]


def _with_ext(
    type_id: int,
    as_id: Optional[AsId],
    raw_config: Optional[ConfigSpec],
    name: Optional[str],
) -> int:
    if as_id:
        res = core.as_id(store, type_id, as_id == "composite")
        if isinstance(res, Err):
            raise ErrorStack(res.value)
        type_id = res.value
    config = serialize_config(raw_config)
    if config:
        res = core.with_config(store, type_id, config)
        if isinstance(res, Err):
            raise ErrorStack(res.value)
        type_id = res.value
    if name:
        res = core.rename_type(store, type_id, name)
        if isinstance(res, Err):
            raise ErrorStack(res.value)
        type_id = res.value
    if name == "ExtendedProfile":
        Log.info(f">>> type#{type_id}; as_id={as_id}; config={config}; name={name}")
    return type_id


class integer(typedef):
    min: Optional[int] = None
    max: Optional[int] = None
    exclusive_minimum: Optional[int] = None
    exclusive_maximum: Optional[int] = None
    multiple_of: Optional[int] = None
    enumeration: Optional[List[int]] = None
    as_id: AsId

    def __init__(
        self,
        *,
        min: Optional[int] = None,
        max: Optional[int] = None,
        exclusive_minimum: Optional[int] = None,
        exclusive_maximum: Optional[int] = None,
        multiple_of: Optional[int] = None,
        enum: Optional[List[int]] = None,
        name: Optional[str] = None,
        config: Optional[ConfigSpec] = None,
        as_id: AsId = False,
    ):
        data = TypeInteger(
            min=min,
            max=max,
            exclusive_minimum=exclusive_minimum,
            exclusive_maximum=exclusive_maximum,
            multiple_of=multiple_of,
            enumeration=enum,
        )
        res = core.integerb(
            store,
            data,
        )
        if isinstance(res, Err):
            raise ErrorStack(res.value)
        super().__init__(_with_ext(res.value, as_id, config, name))
        self.min = min
        self.max = max
        self.exclusive_minimum = exclusive_minimum
        self.exclusive_maximum = exclusive_maximum
        self.multiple_of = multiple_of
        self.enumeration = enum
        self.as_id = as_id

    def id(self, as_id: AsId = True) -> "typedef":  # "integer"
        res = core.as_id(store, self._id, as_id == "composite")
        if isinstance(res, Err):
            raise ErrorStack(res.value)
        return typedef(res.value)


class float(typedef):
    min: Optional[float] = None
    max: Optional[float] = None
    exclusive_minimum: Optional[float] = None
    exclusive_maximum: Optional[float] = None
    multiple_of: Optional[float] = None
    enumeration: Optional[List[float]] = None

    def __init__(
        self,
        *,
        min: Optional[float] = None,
        max: Optional[float] = None,
        exclusive_minimum: Optional[float] = None,
        exclusive_maximum: Optional[float] = None,
        multiple_of: Optional[float] = None,
        enum: Optional[List[float]] = None,
        name: Optional[str] = None,
        config: Optional[ConfigSpec] = None,
    ):
        data = TypeFloat(
            min=og_float(min) if min is not None else None,
            max=og_float(max) if max is not None else None,
            exclusive_minimum=og_float(exclusive_minimum)
            if exclusive_minimum is not None
            else None,
            exclusive_maximum=og_float(exclusive_maximum)
            if exclusive_maximum is not None
            else None,
            multiple_of=og_float(multiple_of) if multiple_of is not None else None,
            enumeration=enum,
        )
        res = core.floatb(
            store,
            data,
        )
        if isinstance(res, Err):
            raise ErrorStack(res.value)
        super().__init__(_with_ext(res.value, None, config, name))
        self.min = min
        self.max = max
        self.exclusive_minimum = exclusive_minimum
        self.exclusive_maximum = exclusive_maximum
        self.multiple_of = multiple_of
        self.enumeration = enum


class boolean(typedef):
    def __init__(
        self, *, name: Optional[str] = None, config: Optional[ConfigSpec] = None
    ):
        res = core.booleanb(store)
        if isinstance(res, Err):
            raise ErrorStack(res.value)
        super().__init__(_with_ext(res.value, None, config, name))


class string(typedef):
    min: Optional[int] = None
    max: Optional[int] = None
    pattern: Optional[str] = None
    format: Optional[str] = None
    enumeration: Optional[List[str]] = None
    as_id: AsId = None

    def __init__(
        self,
        *,
        min: Optional[int] = None,
        max: Optional[int] = None,
        pattern: Optional[str] = None,
        format: Optional[str] = None,
        enum: Optional[List[str]] = None,
        name: Optional[str] = None,
        config: Optional[ConfigSpec] = None,
        as_id: AsId = None,
    ):
        enum_variants = None
        if enum is not None:
            enum_variants = og_list(JsonLib.dumps(variant) for variant in enum)

        data = TypeString(
            min=min, max=max, pattern=pattern, format=format, enumeration=enum_variants
        )

        res = core.stringb(
            store,
            data,
        )
        if isinstance(res, Err):
            raise ErrorStack(res.value)
        super().__init__(_with_ext(res.value, as_id, config, name))
        self.min = min
        self.max = max
        self.pattern = pattern
        self.format = format
        self.enumeration = enum
        self.as_id = as_id

    def id(self, as_id: AsId = True) -> "typedef":  # "integer"
        res = core.as_id(store, self._id, as_id == "composite")
        if isinstance(res, Err):
            raise ErrorStack(res.value)
        return typedef(res.value)


def uuid(
    *,
    config: Optional[ConfigSpec] = None,
    as_id: bool = False,
    name: Optional[str] = None,
) -> string:
    return string(format="uuid", config=config, as_id=as_id, name=name)


def email(
    *,
    config: Optional[Optional[ConfigSpec]] = None,
    as_id: bool = False,
    name: Optional[str] = None,
) -> string:
    return string(format="email", config=config, as_id=as_id, name=name)


def uri() -> string:
    return string(format="uri")


def ean() -> string:
    return string(format="ean")


def path() -> string:
    return string(format="path")


def date() -> string:
    return string(format="date")


def datetime() -> string:
    return string(format="date-time")


def json() -> string:
    return string(format="json")


def enum(
    variants: List[str],
    name: Optional[str] = None,
):
    return string(enum=variants, name=name)


class file(typedef):
    min: Optional[int] = None
    max: Optional[int] = None
    allow: Optional[List[str]]

    def __init__(
        self,
        *,
        min: Optional[int] = None,
        max: Optional[int] = None,
        allow: Optional[List[str]] = None,
        config: Optional[ConfigSpec] = None,
        name: Optional[str] = None,
    ):
        data = TypeFile(
            min=min,
            max=max,
            allow=allow,
        )

        res = core.fileb(
            store,
            data,
        )
        if isinstance(res, Err):
            raise ErrorStack(res.value)

        super().__init__(_with_ext(res.value, None, config, name))
        self.min = min
        self.max = max
        self.allow = allow


class list(typedef):
    items: typedef = None
    min: Optional[int] = None
    max: Optional[int] = None
    unique_items: Optional[bool] = None

    def __init__(
        self,
        items: typedef,
        min: Optional[int] = None,
        max: Optional[int] = None,
        unique_items: Optional[bool] = None,
        name: Optional[str] = None,
        config: Optional[ConfigSpec] = None,
    ):
        data = TypeList(
            of=items._id,
            min=min,
            max=max,
            unique_items=unique_items,
        )

        res = core.listb(
            store,
            data,
        )
        if isinstance(res, Err):
            raise ErrorStack(res.value)
        super().__init__(_with_ext(res.value, None, config, name))
        self.min = min
        self.max = max
        self.items = items
        self.unique_items = unique_items


class optional(typedef):
    item: Optional[typedef] = None
    default_item: Optional[str] = None

    def __init__(
        self,
        item: typedef,
        default_item: Optional[Any] = None,
        name: Optional[str] = None,
        config: Optional[ConfigSpec] = None,
    ):
        data = TypeOptional(
            of=item._id,
            default_item=None if default_item is None else JsonLib.dumps(default_item),
        )

        res = core.optionalb(
            store,
            data,
        )
        if isinstance(res, Err):
            raise ErrorStack(res.value)
        super().__init__(_with_ext(res.value, None, config, name))
        self.item = item
        self.default_item = default_item


class union(typedef):
    variants: List[typedef] = []

    def __init__(
        self,
        variants: List[typedef],
        name: Optional[str] = None,
        config: Optional[ConfigSpec] = None,
    ):
        data = TypeUnion(variants=og_list(map(lambda v: v._id, variants)))

        res = core.unionb(
            store,
            data,
        )
        if isinstance(res, Err):
            raise ErrorStack(res.value)
        super().__init__(_with_ext(res.value, None, config, name))
        self.variants = variants


class either(typedef):
    variants: List[typedef] = []

    def __init__(
        self,
        variants: List[typedef],
        name: Optional[str] = None,
        config: Optional[ConfigSpec] = None,
    ):
        data = TypeEither(variants=og_list(map(lambda v: v._id, variants)))
        res = core.eitherb(
            store,
            data,
        )
        if isinstance(res, Err):
            raise ErrorStack(res.value)
        super().__init__(_with_ext(res.value, None, config, name))
        self.variants = variants


class struct(typedef):
    props: Dict[str, typedef]
    additional_props: bool
    min: Optional[int]
    max: Optional[int]
    enumeration: Optional[List[int]] = None

    def __init__(
        self,
        props: Optional[Dict[str, typedef]] = None,
        *,
        additional_props: bool = False,
        min: Optional[int] = None,
        max: Optional[int] = None,
        name: Optional[str] = None,
        config: Optional[ConfigSpec] = None,
        enum: Optional[List[Dict[str, Any]]] = None,
    ):
        if self.__class__ != struct:  # custom class
            if len(self.__class__.__bases__) > 1:
                raise ErrorStack.from_str(
                    "multiple inheritance is currently not supported"
                )
            (base,) = self.__class__.__bases__
            child_cls = self.__class__
            child_attr = set([i for i in vars(child_cls) if not i.startswith("__")])
            parent_attr = set([i for i in vars(base) if not i.startswith("__")])

            # reserved field check
            reserved_attr = set(vars(struct)).union(vars(typedef))
            common = sorted(reserved_attr.intersection(child_attr))
            if len(common) > 0:
                err_msg = ", ".join(common)
                if len(common) == 1:
                    err_msg += " is a reserved field"
                else:
                    err_msg += " are reserved fields"
                raise ErrorStack(err_msg)
            self_attr = child_attr
            if base != struct:
                # child.props should inherit parent.props
                curr_base = base
                while curr_base != struct:
                    if len(curr_base.__bases__) > 1:
                        raise ErrorStack(
                            "multiple inheritance is currently not supported"
                        )
                    (curr_base,) = curr_base.__bases__
                    fields = set([i for i in vars(curr_base) if not i.startswith("__")])
                    parent_attr = parent_attr.union(fields)
                self_attr = self_attr.union(parent_attr)
            props = {}
            for attr in sorted(self_attr):
                value = getattr(self, attr)
                if isinstance(value, typedef):
                    props[attr] = value

        else:
            props = props or {}

        data = TypeStruct(
            props=og_list((name, tpe._id) for (name, tpe) in props.items()),
            additional_props=additional_props,
            min=min,
            max=max,
            enumeration=[JsonLib.dumps(v) for v in enum] if enum else None,
        )

        res = core.structb(
            store,
            data,
        )
        if isinstance(res, Err):
            raise ErrorStack(res.value)
        super().__init__(_with_ext(res.value, None, config, name))
        self.props = props
        self.enumeration = enum

    def extend(self, props: Dict[str, typedef]):
        return struct(props={**self.props, **props})


ApplyParamObjectNode = Dict[str, "ApplyParamNode"]
ApplyParamArrayNode = List["ApplyParamNode"]
ApplyParamLeafNode = Union[
    ApplyFromArg, ApplyFromStatic, ApplyFromContext, ApplyFromSecret, ApplyFromParent
]

ApplyParamNode = Union[ApplyParamObjectNode, ApplyParamArrayNode, ApplyParamLeafNode]


def serialize_apply_param_node(node: ApplyParamNode) -> Any:
    if isinstance(node, ApplyFromArg):
        return {"source": "arg", "name": node.name}
    if isinstance(node, ApplyFromStatic):
        return {"source": "static", "value_json": JsonLib.dumps(node.value)}
    if isinstance(node, ApplyFromContext):
        return {"source": "context", "key": node.key}
    if isinstance(node, ApplyFromSecret):
        return {"source": "secret", "key": node.key}
    if isinstance(node, ApplyFromParent):
        return {"source": "parent", "type_name": node.type_name}
    if isinstance(node, dict):
        return {
            "type": "object",
            "fields": {k: serialize_apply_param_node(v) for k, v in node.items()},
        }
    if isinstance(node, (og_list, tuple)):
        return {"type": "array", "items": [serialize_apply_param_node(v) for v in node]}

    raise ErrorStack(f"unexpected node type: node={node}")


class func(typedef):
    inp: Union[struct, typedef]
    out: typedef
    mat: Materializer
    rate_calls: bool
    rate_weight: Optional[int]
    parameter_transform: Optional[ParameterTransform]

    def __init__(
        self,
        inp: Union[struct, typedef],
        out: typedef,
        mat: Materializer,
        /,
        rate_calls: bool = False,
        rate_weight: Optional[int] = None,
        parameter_transform: Optional[ParameterTransform] = None,
        type_id: Optional[int] = None,
    ):
        def register():
            data = TypeFunc(
                inp=inp._id,
                out=out._id,
                parameter_transform=parameter_transform,
                mat=mat.id,
                rate_calls=rate_calls,
                rate_weight=rate_weight,
            )
            res = core.funcb(store, data)
            if isinstance(res, Err):
                raise ErrorStack(res.value)
            return res.value

        id = register() if type_id is None else type_id

        super().__init__(id)
        self.inp = inp
        self.out = out
        self.mat = mat
        self.rate_calls = rate_calls
        self.rate_weight = rate_weight
        self.parameter_transform = parameter_transform

    def rate(self, calls: bool = False, weight: Optional[int] = None) -> "func":
        return func(
            self.inp,
            self.out,
            self.mat,
            parameter_transform=self.parameter_transform,
            rate_calls=calls,
            rate_weight=weight,
        )

    def extend(self, props: Dict[str, typedef]):
        res = core.extend_struct(
            store, self.out._id, og_list((k, v._id) for k, v in props.items())
        )
        if isinstance(res, Err):
            raise ErrorStack(res.value)

        out = typedef(res.value)

        return func(
            self.inp,
            out,
            self.mat,
            parameter_transform=self.parameter_transform,
            rate_calls=self.rate_calls,
            rate_weight=self.rate_weight,
        )

    def reduce(self, value: Dict[str, Any]) -> "func":
        reduce_entries = build_reduce_entries(value, [], [])
        reduced_id = wit_utils.reduceb(store, self._id, reduce_entries)

        if isinstance(reduced_id, Err):
            raise ErrorStack(reduced_id.value)

        # TODO typedef(...).as_struct()
        return func(
            self.inp,
            self.out,
            self.mat,
            parameter_transform=self.parameter_transform,
            rate_calls=self.rate_calls,
            rate_weight=self.rate_weight,
            type_id=reduced_id.value,
        )

    def apply(self, value: ApplyParamObjectNode) -> "func":
        serialized = serialize_apply_param_node(value)
        assert isinstance(serialized, dict)
        assert serialized["type"] == "object"
        transform_tree = JsonLib.dumps(serialized["fields"])

        transform_data = core.get_transform_data(store, self.inp._id, transform_tree)
        if isinstance(transform_data, Err):
            import sys

            print(transform_tree, file=sys.stderr)
            raise ErrorStack(transform_data.value)

        return func(
            typedef(transform_data.value.query_input),
            self.out,
            self.mat,
            parameter_transform=transform_data.value.parameter_transform,
            rate_calls=self.rate_calls,
            rate_weight=self.rate_weight,
        )

    @staticmethod
    def from_type_func(
        data: FuncParams, rate_calls: bool = False, rate_weight: Optional[int] = None
    ) -> "func":
        # Note: effect is a just placeholder
        # in the deno frontend, we do not have to fill the effect attribute on materializers
        mat = Materializer(id=data.mat, effect=EffectRead())
        return func(
            typedef(id=data.inp),
            typedef(id=data.out),
            mat,
            rate_calls=rate_calls,
            rate_weight=rate_weight,
            parameter_transform=None,
        )


def gen(
    out: typedef,
    mat: Materializer,
    rate_calls: bool = False,
    rate_weight: Optional[int] = None,
):
    return func(
        struct({}),
        out,
        mat,
        rate_calls=rate_calls,
        rate_weight=rate_weight,
    )
